### ISSUE TITLE
Segmentation tasks

### ISSUE THREAD

**Issue by nabsabraham** [xu-ji/IIC#9]
Hi Xu, thanks for the great code and really novel paper!  I didn't understand the segmentation implementation too well - hoping you can clarify:  
* I want to know how to run your code for a segmentation dataset where I have 4 classes. Would I require 4 heads for this task? 
* Why does `SegmentationNet10aTwoHead` have two output heads with different sized channels? (Talking about `output_k_A` and `output_k_B`). 
* In your experience, does it take a long time to run the `IID` loss for say COCO-Stuff 512 settings? It takes me roughly 7 seconds to compute the loss **per batch** for a 4-class dataset, `batch=32` and `imgsize= 3x128x128` - I'm not sure if this is normal (My setup is GTX1080 Ti, Cuda 10, pytorch 1.0.1)
**EDIT:** The time for loss computation was solved for me by removing the float [here](https://github.com/xu-ji/IIC/blob/fcc13855b5e43bfae7100ce2f337d8438d5ccb48/code/utils/segmentation/IID_losses.py#L60) and by clipping the minimum value to EPS by using the following instead of [this](https://github.com/xu-ji/IIC/blob/fcc13855b5e43bfae7100ce2f337d8438d5ccb48/code/utils/segmentation/IID_losses.py#L71):
```  
p_i_j = torch.clamp(p_i_j, min=EPS, max=10000)
p_i_mat = torch.clamp(p_i_mat, min=EPS, max=10000)
p_j_mat = torch.clamp(p_j_mat, min=EPS, max=10000)
```

**Comment by xu-ji** [comment]
- The two heads are for main output and auxiliary overclustering output. At test time only the main head is used, the overclustering head is ignored (section 3.2 in the paper, auxiliary overclustering). Each of these heads has sub-heads, which are just multiple instantiations with random initialisations (can be just 1, in the simple case). Each sub-head has `c` clusters (the same for all-subheads of the head). So you would have a main head with `c = 4` and the auxiliary head with some `c > 4` (even the auxiliary head is optional, but performance improves a lot if it is used.) The former is `output_k_B` and the latter is `output_k_A` in the code. These are the number of output channels in the network, just like in supervised classification. They are different because the number of clusters for each head is different - one is ground truth number of clusters, the other is larger for overclustering.
- COCO-Stuff 512 took about 2 seconds per batch.

**Comment by nabsabraham** [comment]
Thank you for the clarification! I re-read version 1 of your paper and concepts became more clear. I just have one last q if you don't mind - I read somewhere in your work that the ground truth has to be one-hot encoded - so at [this](https://github.com/xu-ji/IIC/blob/fcc13855b5e43bfae7100ce2f337d8438d5ccb48/code/scripts/segmentation/segmentation_twohead.py#L303) point in your code, shouldn't you be doing the masking for every single class' ground truth maps?

**Comment by xu-ji** [comment]
The mask in that line is used for ignoring predictions for non-Stuff pixels when computing the loss. It simply says which pixels are part of the dataset and which aren't. The ground truth labels are not used, since this is unsupervised learning. Except when evaluating performance.

**Comment by RyanCV** [comment]
@nabsabraham @xu-ji  Could you guys give more detail about how to generate my own dataset? What is the format?
Thanks.
